### PR TITLE
Fix check canonical signature

### PR DESCRIPTION
**Pull request by afk11**
Rewrote encode_signature() and is_canonical_signature(). These are tested against vectors from the Satoshi client.

### CODE REVIEW

**Comment by rubensayshi** [comment]
looks good.

 - I think we should start throwing exceptions more often instead of silently failing and returning FALSE.
 - should we still do an `is_canonical_signature` after signing a TX? or are we going to asume it's always correct?

**Comment by btcdrak** [comment]
needs rebase.

**Comment by rubensayshi** [comment]
I'm trying to get a randomK that results in a non canonical sig, but so far after 1mil loops I haven't gotten one yet ...
Would be nice to have a test vector that fails ...

**Comment by afk11** [comment]
Sweet, that's good news! When I was debugging that code it would normally happen after a few hundred. I went for dealing with a packed byte string in the new functions, hopefully it'll remove any room for error..

Yeah, I would love this too, for the r and s failures. I'll try take a look through https://github.com/bitcoin/bitcoin or https://github.com/bitcoin/libsecp256k1 for something. 

I was thinking it could be good to throw an exception for non canonical signatures, though may warrant a version bump?

**Comment by rubensayshi** [comment]
imo let's throw an exception, if the problems persist then users can catch it and retry.
we can merge it together with the other PRs (https://github.com/Bit-Wasp/bitcoin-lib-php/pull/48 and https://github.com/Bit-Wasp/bitcoin-lib-php/pull/33) which break BC.

it would be really great to have some test vectors with this though, specially one that gives a non canonical signature with the old code and now works ... but I can't seem to get one, even after running a loop for a few days :/

File: src/RawTransaction.php [hunk 1]
```diff
@@ -1208,29 +1208,35 @@ public static function extract_input_signatures_p2sh($input, $message_hash, $scr
...
+            .hex2bin(dechex(strlen($rBin)))
+            .$rBin
+            .hex2bin('02')
+            .hex2bin(dechex(strlen($sBin)))
+            .$sBin;
 
+        $outer = hex2bin('30')
+            .hex2bin(dechex(strlen($inner)))
+            .$inner
+            .hex2bin('01');
```
Comment (rubensayshi) [inline src/RawTransaction.php]: @afk11 @btcdrak I think this 01 at the end should not be here, bitcore and bitcoinjs don't add this and they both see this our signatures as non canonical ...

you also seem to compensate for this suffix in various places, such as doing -3 on the length byte when checking if the sig is canonical, but the length byte should be 1 (length byte R) + len(R) + 1 (length byte S) + len(S).
so the length of the sig should be length byte + 1 (sequence) + 1 (length byte)

**Comment by afk11** [comment]
This came up on the mailing list too I think - so the byte is tacked on, but has nothing to do with DER encoding. It indicates the hashtype of the input signature, since there are different ways to sign your input. 

The data your signing (per input) is a mangled version of the transaction committing only certain details in the signature of an input. The procedure depends on the bitmask held in that one byte.

Unfortunately, the library doesn't support all of these hashtypes, just sighash all. It would probably fail to verify anything but this hashtype. 

The check for canonical signatures was pretty much taken from here: 
https://github.com/bitcoin/bitcoin/blob/b01a435c34f0669affece2b3d90f92c347588d69/src/script/interpreter.cpp#L97 along with tests for it's failures. 

So the only way you could fail is_canonical_signature because of that last byte is the length checks. But, a side effect of getting the byte wrong means you come up with the wrong hash during signature verification.

**Comment by rubensayshi** [comment]
okay, at least that takes away the worries ;-)  
I'll go dig deeper and figure out where those nodejs libs seem to trim off that byte, just out of curiousity :D

I still haven't managed to get a test vector that fails, also haven't had a failure in production for a few days now ...  
I think I'll modify my version of bitcoin-lib temporarly to include the R value in the exception message when it fails, that way I can log it and hopefully we'll have a test vector we can work with soon...

**Comment by rubensayshi** [comment]
I've also noticed that bitcore and bitcoinjs-lib don't validate their 'work', so no `is_canonical_signature` and no `validate_signed_transaction`.  
Since I wanted to benchmark some stuff I removed that stuff from RawTransaction, that's about an ~60% performance improvement...

If we would be certain that we never produce non canonical signatures then we could also stop doing those checks :)

**Comment by afk11** [comment]
Cool, that would be neat! I'm not sure how extensively they validate it - I know Copay fell victim to an exploit against SIGHASH types back in August. There seems to be some support for it there anyway. 

It's a reasonable question as to whether it should be stored in a Signature class, or just stored elsewhere... Had the same question in my other lib.. Made that repo public now!

I am pretty sure that it is producing correctly encoded signatures now. If you've been running only code consistency tests then it's good nothings come up - as it can detect failing cases. If bitcoind has been verifying those transactions as correctly signed in production then even better :)

The main reason the checking work is thereis mainly for checking if the transaction is fully signed, as bitcoind will tell you through complete = bool.. We can take some shortcuts I guess, perhaps if it finds a canonical signature, then it counts that as a signature, and tracks added signatures to see if it's complete?

**Comment by afk11** [comment]
Whoop, they passed!

**Comment by rubensayshi** [comment]
ouch @afk11 when you resolved the merge conflicts on this one you undid my work on replacing the `return false` stuff with `throw \InvalidArgumentException` :(

I'll do another PR to patch it ;-)

**Comment by afk11** [comment]
Damnit yeah, I meant to get to that.. Merged that one first and eventually lost it through conflicts

**Comment by rubensayshi** [comment]
fixed in https://github.com/Bit-Wasp/bitcoin-lib-php/pull/65

**Pull request merged**